File: proverb/Proverb.fs
module Proverb

let recite (input: string list): string list = 
    let reciter acc x = 
        match acc with 
        | "" -> ("", x) 
        | _ -> (sprintf "For want of a %s the %s was lost." acc x, x)
    match input with
    | [] -> []
    | _ -> input |> List.mapFold reciter "" |> fst |> List.skip 1 |> (fun x -> x @ [sprintf "And all for the want of a %s." input.[0]])

File: csharp/proverb/Proverb.cs
using System;
using System.Linq;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    { 
        if (subjects.Length == 0)
            return new string[0];

        return subjects
            .Zip(subjects.Skip(1), (first, second) => (first, second))
            .Select(pair => $"For want of a {pair.first} the {pair.second} was lost.")
            .Append($"And all for the want of a {subjects[0]}.").ToArray();
    }
}